libzfs/src/lib.rs

```diff
@@ -0,0 +1,326 @@
+            })
+        }
+        ref x if x == sys::VDEV_TYPE_FILE => {
+            let path = tree.lookup_string("path")?;
+
+            Ok(VDev::File { path })
+        }
+        ref x if x == sys::VDEV_TYPE_MIRROR => {
+            let children = get_children(tree)?;
+
+            Ok(VDev::Mirror { children })
+        }
+        ref x if x == sys::VDEV_TYPE_RAIDZ => {
+            let children = get_children(tree)?;
+
+            Ok(VDev::RaidZ { children })
+        }
+        ref x if x == sys::VDEV_TYPE_REPLACING => {
+            let children = get_children(tree)?;
+
+            Ok(VDev::Replacing { children })
+        }
+        ref x if x == sys::VDEV_TYPE_SPARE => {
+            let children = get_children(tree)?;
+
+            Ok(VDev::Spare { children })
+        }
+        ref x if x == sys::VDEV_TYPE_ROOT => {
+            let children = get_children(tree)?;
+
+            Ok(VDev::Root { children })
+        }
+        _ => Err(Error::new(ErrorKind::NotFound, "hit unknown vdev type")),
+    }
+}
+
+impl Drop for Zpool {
+    fn drop(&mut self) {
+        unsafe { sys::zpool_close(self.raw) }
+    }
+}
+
+pub struct Zfs {
+    raw: *mut sys::libzfs_handle_t,
+}
+
+impl Zfs {
+    pub fn new() -> Zfs {
+        Zfs {
+            raw: unsafe { sys::libzfs_init() },
+        }
+    }
+    pub fn find_importable_pools(&mut self) -> nvpair::NvList {
+        unsafe {
+            sys::thread_init();
+            let x = sys::zpool_find_import(self.raw, 0, ptr::null_mut());
+            sys::thread_fini();
+
+            nvpair::NvList::from_ptr(x, true)
+        }
+    }
+    pub fn import_all(&mut self, nvl: &nvpair::NvList) -> io::Result<Vec<()>> {
+        nvl.iter().map(|x| {
+            let nvl2 = x.value_nv_list()?;
+
+            let code =
+                unsafe { sys::zpool_import(self.raw, nvl2.as_ptr(), ptr::null(), ptr::null_mut()) };
+
+            match code {
+                0 => Ok(()),
+                x => Err(io::Error::from_raw_os_error(x)),
+            }
+        }).collect()
+    }
+    pub fn export_all(&mut self, pools: &Vec<Zpool>) -> io::Result<Vec<()>> {
+        pools
+            .iter()
+            .map(|x| {
+                let code =
+                    unsafe { sys::zpool_export(x.raw, sys::boolean_B_FALSE, ptr::null_mut()) };
+
+                match code {
+                    0 => Ok(()),
+                    x => Err(io::Error::from_raw_os_error(x)),
+                }
+            })
+            .collect()
+    }
+    pub fn get_imported_pools(&mut self) -> io::Result<Vec<Zpool>> {
+        unsafe extern "C" fn callback(
+            handle: *mut sys::zpool_handle_t,
+            state: *mut c_void,
+        ) -> c_int {
+            let state = &mut *(state as *mut Vec<Zpool>);
+
+            state.push(Zpool { raw: handle });
+
+            0
+        }
+        let mut state: Vec<Zpool> = Vec::new();
+        let state_ptr: *mut c_void = &mut state as *mut _ as *mut c_void;
```

Comment:
asked before and I'm not sure you gave an answer, why the second cast?

Reply:
It takes two casts to get the vec into a `c_void` type that's acceptable to pass into a c callback.